sql: refactor instrumentation of query execution

RaduBerinde opened the pull request:
Note: this PR covers only the two top commits; the others are #56262.

#### sql: refactor instrumentation of query execution

This change separates the code around setting up tracing and
collecting bundles into a new instrumentationHelper. The intention is
that more related logic (like collecting plans for the UI) will be
incorporated.

Release note: None

#### sql: move more logic into instrumentationHelper

This commit moves the logic around collecting explain plans (for
bundles and for UI) into `instrumentationHelper`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56286)
<!-- Reviewable:end -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice refactor

> Reviewed 4 of 4 files at r1, 6 of 6 files at r2, 5 of 5 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56286#-:-MLNR45k41F-b8CwnAWu:bjm753q)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @RaduBerinde)

___
*[pkg/sql/instrumentation.go, line 195 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56286#-MLNOzIF8QGiMO00GmYb:-MLNOzIF8QGiMO00GmYc:b-agxbth) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf32f545a5c882c4c365b4497979988dcd7fa2c/pkg/sql/instrumentation.go#L195)):*
> ```Go
> 
> // SetDiscardRows should be called when we want to discard rows for a
> // non-ANALYZE statement (via EXECUTE .. DISCARD ROWS.
> ```

nit: parenthesis is not closed


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: move more logic into instrumentationHelper

This commit moves the logic around collecting explain plans (for
bundles and for UI) into `instrumentationHelper`.

Release note: None

RaduBerinde committed:
sql: refactor instrumentation of query execution

This change separates the code around setting up tracing and
collecting bundles into a new instrumentationHelper. The intention is
that more related logic (like collecting plans for the UI) will be
incorporated.

Release note: None

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56286#-:-MLOrayEAsYSErci41c0:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @asubiotto)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
